graphs/coloring/sync_goldberg_plotkin_shannon/gps.go

```diff
@@ -0,0 +1,523 @@
+package sync_goldberg_plotkin_shannon
+import(
+	"github.com/krzysztof-turowski/distributed-framework/lib"
+	"fmt"
+	"encoding/json"
+	"log"
+	"math"
+)
+
+
+func Run(n int, p float64) (int, int) {
+	nodes, synchronizer := lib.BuildSynchronizedRandomGraph(n, p)
+	delta := 0
+	id := 0
+	for _,node := range nodes {
+		delta = max(delta,node.GetOutChannelsCount()) //each node must know max degree in graph
+	}
+	for _, node := range nodes {
+		log.Println("Node", node.GetIndex(), "about to run")
+		go run(node,delta,id)
+		id++
+	}
+	synchronizer.Synchronize(3)
+	check(nodes)
+	return synchronizer.GetStats()
+}
+
+func check(nodes []lib.Node) {
+	for _, node := range nodes {
+		our_color := getState(node).Color
+		if(our_color > getState(node).Graph_size) {
+			panic("too many colors!")
+		}
+		for _, neighbor := range node.GetOutNeighbors() {
+			if getState(neighbor).Color == our_color {
+				panic("two adjacent nodes have the same color!")
+			}
+		}
+		
+	}
+}
+
+
+
+func run(node lib.Node,delta int,id int) {
+	node.StartProcessing()
+	
+	finish := initialize(node,delta,id)
+	node.FinishProcessing(finish)
+
+	for round := 1; !finish; round++ {
+		node.StartProcessing()
+		finish = process(node)
+		
+		node.FinishProcessing(finish)
+		
+	}
+}
+func initialize(node lib.Node, delta int,id int) bool {
+	
+	setState(node, makeState(node.GetOutChannelsCount(), delta,node.GetSize(),id))
+	return false
+}
+
+func process(node lib.Node) bool {
+	s := getState(node)
+	switch s.Stage {
+	case pseudoforest_creation:
+		s.choose_edge_for_next_pseudoforest(node)
+		setState(node,s)
+	case coloring_pseudoforests_six_colors:
+		s.reduce_colors_logarythmically(node)
+		setState(node,s)
+	case coloring_pseudoforests_with_three_colors:
+		s.remove_additional_three_colors(node)
+		setState(node,s)
+	case recoloring_graph:
+		s.final_recoloring(node)
+		setState(node,s)
+	case colored:
+		return s.Done
+	default:
+		panic(fmt.Sprint("Invalid stage ", s.Stage))
+	}
+	return s.Done
+}
+
+//main algorithm 
+func (s *state)final_recoloring(node lib.Node) {
+	if(s.Done) {
+			
+		return
+	}	
+	s.Iteration++
+	if(s.Iteration == 1) {
+		if _,exists := s.Pseudoforests_in[s.Current_pseudoforest];exists {
+			
+			s.Edges_left_for_taking = append(s.Edges_left_for_taking, s.Pseudoforests_in[s.Current_pseudoforest]...)
+				if(s.Pseudoforests_out[s.Current_pseudoforest] != -1) {
+					s.Edges_left_for_taking = append(s.Edges_left_for_taking, s.Pseudoforests_out[s.Current_pseudoforest])
+				}
+		}
+	}
+		if(s.Iteration > s.Delta) {
+			s.Iteration = 1
+			s.Current_pseudoforest--
+			if(s.Current_pseudoforest < 0) {
+				s.Current_pseudoforest = s.Delta
+				s.K++
+				
+			} else {
+				
+			}
+			if(s.K == 3) {
+				s.Done = true
+				s.Stage = colored
+			}
+			
+			return 
+		}
+	
+		if _,exists := s.Pseudoforests_in[s.Current_pseudoforest];!exists {
+			return
+			
+		}
```

Comment:
Please remove empty line, add space after comma (before exists) etc. - in general please run gofmt -l -s -w FILE` on files you created